🎳1단계 - 질문 삭제하기 기능 리팩토링

boradol opened the pull request:
안녕하세요.
리뷰 잘 부탁드립니다.🙏🏻

boradol committed:
docs(README.md): step 1. 기능 명세.

boradol committed:
docs(QnaService): Service 에서 도메인으로 비즈니스 이동할 코드 삭제

boradol committed:
docs(Question, Answer): 도메인에 비즈니스 로직 이동.

1. delete true 변환
2. DeleteHistory 내역 추가시킴
3. validate 구현.
4. 불필요한 getter/setter 코드 삭제.

boradol committed:
docs(Answers): 일급컬렉션 생성.

1. @Embedded, @Embeddable 사용.

seondongpyo commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,40 +1,42 @@
     private boolean deleted = false;
 
-    public Answer() {
-    }
```
JPA를 사용할 때 엔티티 클래스에는 기본 생성자가 필요하니 삭제하지 않으시면 좋을 것 같습니다.
https://hyeonic.tistory.com/191

seondongpyo commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.List;
+
+@Embeddable
+public class Answers {
```
일급 콜렉션 👍

seondongpyo commented on src/main/java/qna/service/QnAService.java:
```diff
@@ -35,24 +35,6 @@ public Question findQuestionById(Long id) {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        deleteHistoryService.saveAll(question.delete(loginUser));
```
👍

seondongpyo commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,28 @@
+    }
+
+    @Test
+    @DisplayName("본인이 답변 작성자가 아면 삭제 실패하고 에러 발생 한다.")
```
```suggestion
    @DisplayName("본인이 답변 작성자가 아니면 삭제 실패하고 에러 발생 한다.")
```

seondongpyo commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,41 @@
     public static final Question Q2 = new Question("title2", "contents2").writeBy(UserTest.SANJIGI);
+
+    @Test
+    @DisplayName("질문작성자가 답변이 없는 질문을 삭제하면, 삭제상태가 되고 삭제 내역을 반환해야 한다.")
```
질문 작성자가 답변이 있는 질문을 삭제하는 경우도 테스트해보면 어떨까요?

seondongpyo commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,41 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
 public class QuestionTest {
```
다음 요구사항에 대한 테스트도 추가해보면 좋을 것 같아요.
`질문을 삭제할 때 답변 또한 삭제해야 하며, 답변의 삭제 또한 삭제 상태(deleted)를 변경한다.`

boradol replied:
위의 테스트와 동일한 곳에서 처리하도록 하겠습니다.

seondongpyo commented on src/test/java/qna/domain/AnswersTest.java:
```diff
@@ -0,0 +1,32 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class AnswersTest {
+    private static final Answers answers = new Answers();
```
전체적으로 단위 테스트들을 실행할 때 `deleteAnswersSuccessTest()`, `deleteAnswersFailByIsNotWriter()` 순으로 
실행이 된다면 다행이지만, 테스트 실행 순서는 개발자가 지정해주지 않으면 기본적으로 보장이 되지 않습니다.
그렇기 때문에 다음과 같이 순서가 바뀌면 테스트가 통과하지 않는 경우도 발생합니다.
단위 테스트를 순서에 상관없이 독립적으로 실행이 가능하게끔 작성해보는 건 어떨까요?

![image](https://user-images.githubusercontent.com/64854054/199656801-05c95b75-725b-4df2-b44e-abccb0b4a9d3.png)

seondongpyo reviewed (changes requested):
안녕하세요, 보라님. 볼링 점수판 미션을 함께하게 된 표선동이라고 합니다 🙇
1단계 리팩토링 미션을 잘 작성해주셨네요 👍
테스트와 관련하여 몇 가지 코멘트 드렸는데 확인 후 다시 리뷰 요청을 부탁드리곘습니다.
궁금하시거나 나누고 싶은 의견 있으시면 편하게 코멘트나 Slack DM 주시길 바랄게요.
조금만 더 힘내주세요 🔥

boradol committed:
docs(README.md): 리뷰 요구사항 명세.

boradol committed:
refactor(Answer, Question): 기본생성자 살리기.

boradol committed:
refactor(Question): 질문 삭제 테스트 추가

1. 질문의 답변이 질문작성자만 답변하였을 때 삭제 성공
2. 질문과 답변은 delete -> true
3. 삭제 내역은 질문과 답변 각각 1개씩 해서 총 2개 생김.

boradol committed:
refactor(Answers): 테스트 끼리 로직에 영향받지 않게 테스트 구현.

boradol committed:
refactor(AnswerTest): 오탈자 수정.

seondongpyo reviewed (approved):
피드백을 잘 반영해주셨네요 👍 
크게 드릴 말씀이 없어서 바로 merge 하겠습니다.
다음 단계도 힘내시고, 즐거운 주말 되세요 🔥
